### PR TITLE
Support subteams that have subteams

### LINKED ISSUE #1854: Support for subteams that have subteams
### Summary

Add support for subteams that have subteams (which rust-lang/team#1046 introduces) on the governance pages.
Subteams of subteams are to be laid out flat below their parent team on the respective root team page.
Although it's lacking practical relevance, subteams of arbitrary depth should be supported.

### Motivation

In rust-lang/team#1040, we'd like to create the rustdoc subteam *rustdoc-contributors*. However rustdoc itself is already a subteam (namely of devtools) and we currently lack support for subsubteams (and so forth) in the teams repo and on the website.
It's easy to update the teams repo to make it understand deeply nested teams (just one check has to be removed). I'm pretty sure it's relatively straightforward to update the website, too. The creation of *rustdoc-contributors* isn't necessarily blocked on website support (the team just wouldn't show up[^1] if we were to add the team right now) but it'd certainly be nice to have.

### Drawbacks

Laying out sub<sup>n</sup>teams (n≥2) flat below their parent team might be confusing for website visitors who expect the list of teams to only contain “direct descendants”.

### Rationale and alternatives

I think this design is the most straightforward one to implement.

Alternatively, we could lay out those subteams hierarchically where clicking on the section header of a subteam leads to another subpage and so on. This however would introduce an inconsistency since section headers of subteams without further subteams would be link anchors contrary to those with child teams.

### Maintenance

This would only add new *logic* to the website internals but not any new *data* that would need to be maintained because everything would be pulled from the `rust_team_data` API. As this isn't a big feature, this can be maintained by the one who already maintains the team-related parts of the website.

### Unresolved Questions

* Should we add a paragraph or similar to the section of sub<sup>n</sup>teams noting that team is actually a subteam of another subteam (where we specify the relevant parent team) or would that be too bloated?
* Should we lay out sub<sup>n</sup>teams flat or hierarchically (see Alternatives)?
* Should there be a maximum supported depth (this limit can be enforced by the CI of the teams repo)?

[^1]: I've confirmed this by building the website locally with a patched teams repo & a patched `rust_team_data` static API.

### PR DESCRIPTION
**Pull request by fmease**
Fixes #1854.

Renders subteams of arbitrary depth.

The the "flat design" doesn't allow for subteams to have working groups or project groups since the WGs and PGs are rendered separately below all other teams and therefore can't possibly be placed below their parent team. They simply won't be rendered. I'm gonna update rust-lang/team#1046 to disallow such groups. 

CC @GuillaumeGomez

subsubteam

### CODE REVIEW

File: src/teams.rs [hunk 1]
```diff
@@ -106,17 +106,36 @@ impl Data {
...
+                            .iter()
+                            .filter(|team| team.subteam_of.as_ref() == Some(&subteam.name))
+                            .filter(|team| matches!(team.kind, TeamKind::Team))
+                            .filter(|team| team.website_data.is_some())
+                            .cloned(),
+                    );
+                }
+                TeamKind::WorkingGroup => wgs.push(subteam.clone()),
+                TeamKind::ProjectGroup => project_groups.push(subteam.clone()),
                 _ => {}
```
Comment (GuillaumeGomez) [inline src/teams.rs:127]: Can't we put this into a function which can eventually call itself? Like that we wouldn't have trouble if have super nested subteams/subgroups.

Comment (fmease) [reply src/teams.rs:127]: I can try to impl an recursive or iterative approach. Earlier I've attempted sth. like that it but since the teams are stored in a `Vec`, not a `HashMap` or graph, I had to linearly look through the `Vec` again and again, yielding a bad algorithmic complexity. Let me try again

File: src/teams.rs [hunk 2]
```diff
@@ -101,25 +102,78 @@ impl Data {
...
+            main_team: &str,
+            subteams: &[Team],
+        ) {
+            let name = team.map(|team| team.name.as_str()).or(Some(main_team));
+
+            if let Some(team) = team {
+                result.push(team.clone());
+            }
+
+            for subteam in subteams {
```
Comment (fmease) [inline src/teams.rs:170]: Not super happy about this loop inside this recursive function. Complexity could be better but input size should be relatively low. Let me know if there's a nice way to improve the algorithmic complexity.

**Review by fmease — COMMENTED** [review]
Okay, I now support subteams of arbitrary depth. Should I add a `#[test] fn`?

File: src/teams.rs [hunk 3]
```diff
@@ -101,25 +102,78 @@ impl Data {
...
         project_groups.sort_by_key(|team| Reverse(team.website_data.as_ref().unwrap().weight));
 
+        // Lay out subteams according to their hierarchy.
+        // Superteams come first and subteams come after, recursively.
+        // The ordering of sibling teams is unaffected, they remain sorted first by weight, then
+        // alphabetically if we assume that the Team API serves the teams alphabetically.
+        let mut subteams = Vec::with_capacity(raw_subteams.len());
+        lay_out_subteams_hierarchically(&mut subteams, None, &main_team.name, &raw_subteams);
+
+        fn lay_out_subteams_hierarchically<'a>(
```
Comment (fmease) [inline src/teams.rs:158]: If you're wondering why we need to re-lay out the subteams at all after having fetched all direct and transitive subteams, it's because `raw_subteams` is not necessarily ordered hierarchically.

This might sound obvious but in the case of *rustdoc* and *rustdoc-contributors*, the original ordering is actually correct (coincidentally) due to the naming of the two teams (them being sorted alphabetically implies being sorted hierarchically as they share the same prefix and the latter is longer).

However, in general this is not the case (the names might not necessarily reflect the hierarchy). Especially if we add *weights* to the equation: We don't want to place subsubteams with weight 1000 at the top of the page above everything else (inclusing their parent team!). The weight should be relative to the level / rank / tier. `lay_out_subteams_hierarchically` guarantees that.

**Comment by fmease** [comment]
@Manishearth, do you like this approach of laying out subsubteams (etc.)?

Namely arranging them flat on the page of a main team? This design doesn't introduce any new pages or clutter but it doesn't allow subteams to have WGs and PGs (see PR description). OTOH, it might be confusing to the reader since the hierarchy is not obvious.

Alternatively, we could create subpages for subsubteams (etc.) but I don't know how easy it would be to implement and I'm not sure how we would link to them because at the moment the section headers of subteams and WGs are already *link anchors*.

**Comment by Manishearth** [comment]
I think that's fine!

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
This should show up on www-staging.rust-lang.org in a bit, and after that we should check and see if it works (and then make a PR from main to `deploy` to deploy)

**Comment by fmease** [comment]
> This should show up on www-staging.rust-lang.org in a bit, and after that we should check and see if it works (and then make a PR from main to deploy to deploy)

The website shouldn't have changed since neither rust-lang/team#1046 nor rust-lang/team#1040 has been merged yet.

**Comment by Manishearth** [comment]
Yes, but this code will be running and we need to verify it hasn't broken everything before deploy :smile:
